session.go

```diff
@@ -68,6 +68,83 @@ func newSession(sessionKey int64, clientKey string, conn *websocket.Conn) *Sessi
 	}
 }
 
+// addConnection safely registers a new connection in the connections map
+func (s *Session) addConnection(connID int64, conn *connection) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.conns[connID] = conn
+	if PrintTunnelData {
+		logrus.Debugf("CONNECTIONS %d %d", s.sessionKey, len(s.conns))
+	}
+}
+
+// removeConnection safely removes a connection by ID, returning the connection object
+func (s *Session) removeConnection(connID int64) *connection {
+	s.Lock()
+	defer s.Unlock()
+
+	conn := s.conns[connID]
+	delete(s.conns, connID)
+	if PrintTunnelData {
+		defer logrus.Debugf("CONNECTIONS %d %d", s.sessionKey, len(s.conns))
+	}
+	return conn
+}
+
+// getConnection retrieves a connection by ID
+func (s *Session) getConnection(connID int64) *connection {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.conns[connID]
+}
+
+// activeConnectionIDs returns an ordered list of IDs for the currently active connections
+func (s *Session) activeConnectionIDs() []int64 {
+	s.RLock()
+	defer s.RUnlock()
+
+	res := make([]int64, 0, len(s.conns))
+	for id := range s.conns {
+		res = append(res, id)
+	}
+	sort.Slice(res, func(i, j int) bool { return res[i] < res[j] })
+	return res
+}
+
+// addSessionKey registers a new session key for a given client key
+func (s *Session) addSessionKey(clientKey string, sessionKey int) {
+	s.Lock()
+	defer s.Unlock()
+
+	keys := s.remoteClientKeys[clientKey]
+	if keys == nil {
+		keys = map[int]bool{}
+		s.remoteClientKeys[clientKey] = keys
+	}
+	keys[sessionKey] = true
+}
+
+// removeSessionKey removes a specific session key for a client key
+func (s *Session) removeSessionKey(clientKey string, sessionKey int) {
+	s.Lock()
+	defer s.Unlock()
+
+	keys := s.remoteClientKeys[clientKey]
+	delete(keys, sessionKey)
+	if len(keys) == 0 {
+		delete(s.remoteClientKeys, clientKey)
+	}
+}
+
+// getSessionKeys retrieves all session keys for a given client key
+func (s *Session) getSessionKeys(clientKey string) map[int]bool {
+	s.Lock()
+	defer s.Unlock()
```

Comment:
```suggestion
	s.RLock()
	defer s.RUnlock()
```

Reply:
Oops, I missed this one, thank you. Good catch!